### DISCUSSION TITLE
Kalman Core update with yaw error

### DISCUSSION THREAD

**Discussion by yneuffer** [bitcraze/discussions#870] [Q&A]
Could one use the function kalmanCoreUpdateWithYawError to integrate a measured change in yaw angle from one timestamp to another? Its not quite clear to me whether the state D2 in the kalman S matrix is an just an uncertainty or an actual yaw angle. What would one use as the parameters of the function for a certain measured change in angle, and what does error->yawError represent? Or would you recommend a different approach?

Thanks!

**Comment by krichardsson** [comment]
The way I understand this is that the member [`q[]`](https://github.com/bitcraze/crazyflie-firmware/blob/e56b7fddf6e6ac9e10d9bacfffb4c4bfcd629219/src/modules/interface/kalman_core/kalman_core.h#L83) holds the current attitude as a quaternion. 
`D0` - `D2` in the [`S`](https://github.com/bitcraze/crazyflie-firmware/blob/e56b7fddf6e6ac9e10d9bacfffb4c4bfcd629219/src/modules/interface/kalman_core/kalman_core.h#L78) matrix contains the attitude error (delta) compared to the attitude in q, which means D2 contains an angle.

This functionality is used today in[ the yaw estimation for the lighthouse positioning system](https://github.com/bitcraze/crazyflie-firmware/blob/e56b7fddf6e6ac9e10d9bacfffb4c4bfcd629219/src/modules/src/lighthouse/lighthouse_position_est.c#L447-L453), but it is a bit dirty and should be moved into the kalman filter at some point.

> what does error->yawError represent

The delta of your measured yaw and the yaw from the current estimated yaw (in `q`)

You can also look at [mm_pose.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/kalman_core/mm_pose.c) for inspiration. If you are interested in setting the attitude in absolute numbers you could maybe extract the attitude part from `kalmanCoreUpdateWithPose()`
